osu.Game/Screens/SelectV2/BeatmapCarouselFilterGrouping.cs

```diff
@@ -36,137 +35,115 @@ public BeatmapCarouselFilterGrouping(Func<FilterCriteria> getCriteria)
             this.getCriteria = getCriteria;
         }
 
-        public async Task<IEnumerable<CarouselItem>> Run(IEnumerable<CarouselItem> items, CancellationToken cancellationToken) => await Task.Run(() =>
+        public async Task<IEnumerable<CarouselItem>> Run(IEnumerable<CarouselItem> items, CancellationToken cancellationToken)
         {
-            setItems.Clear();
-            groupItems.Clear();
-
-            var criteria = getCriteria();
-            var newItems = new List<CarouselItem>(items.Count());
-
-            // Add criteria groups.
-            switch (criteria.Group)
+            return await Task.Run(() =>
             {
-                default:
-                    BeatmapSetsGroupedTogether = true;
-                    newItems.AddRange(items);
-                    break;
+                setItems.Clear();
+                groupItems.Clear();
 
-                case GroupMode.Artist:
-                    BeatmapSetsGroupedTogether = true;
-                    char groupChar = (char)0;
-
-                    foreach (var item in items)
-                    {
-                        cancellationToken.ThrowIfCancellationRequested();
+                var criteria = getCriteria();
+                var newItems = new List<CarouselItem>();
 
-                        var b = (BeatmapInfo)item.Model;
+                BeatmapInfo? lastBeatmap = null;
+                GroupDefinition? lastGroup = null;
 
-                        char beatmapFirstChar = char.ToUpperInvariant(b.Metadata.Artist[0]);
+                HashSet<CarouselItem>? groupRefItems = null;
+                HashSet<CarouselItem>? setRefItems = null;
 
-                        if (beatmapFirstChar > groupChar)
-                        {
-                            groupChar = beatmapFirstChar;
-                            var groupDefinition = new GroupDefinition($"{groupChar}");
-                            var groupItem = new CarouselItem(groupDefinition) { DrawHeight = GroupPanel.HEIGHT };
-
-                            newItems.Add(groupItem);
-                            groupItems[groupDefinition] = new HashSet<CarouselItem> { groupItem };
-                        }
+                switch (criteria.Group)
+                {
+                    default:
+                        BeatmapSetsGroupedTogether = true;
+                        break;
 
-                        newItems.Add(item);
-                    }
+                    case GroupMode.Difficulty:
+                        BeatmapSetsGroupedTogether = false;
+                        break;
```

Comment:
This block could be a plain `if` now.

Reply:
It could yeah. Just not sure how this is going to handle going forward. It's definitely not the final iteration as users expect more nuanced behaviour.

I'll change it for now, though.